### PR TITLE
Check for simulator runtime in flutter doctor

### PR DESCRIPTION
**Pull request by vashworth**
Starting in Xcode 15, the simulator is no longer included in Xcode and must be downloaded and installed separately. This adds a validation to `flutter doctor` to warn when the needed simulator runtime is missing.

Validation message looks like:
```
[!] Xcode - develop for iOS and macOS (Xcode 15.0)
    ! iOS 17.0 Simulator not installed; this may be necessary for iOS and macOS development.
      To download and install the platform, open Xcode, select Xcode > Settings > Platforms,
      and click the GET button for the required platform.

      For more information, please visit:
        https://developer.apple.com/documentation/xcode/installing-additional-simulator-runtimes
```

It may also show an error like this when something goes wrong when checking for the simulator:
```
[!] Xcode - develop for iOS and macOS (Xcode 15.0)
    ✗ Unable to find the iPhone Simulator SDK.
```

Note: I'm unsure of in the future if the SDK and the simulator runtime will need to match the exact version or just the major. For now, it only checks against the major version.

Part 3 of https://github.com/flutter/flutter/issues/129558.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by vashworth** [comment]
Google testing is failing, but want to get a LGTM before I submit a g3fix in case I need to change anything

File: packages/flutter_tools/lib/src/ios/simulators.dart [hunk 1]
```diff
@@ -624,6 +673,54 @@ class IOSSimulator extends Device {
...
+  //       "bundlePath" : "\/Applications\/Xcode.app\/Contents\/Developer\/Platforms\/iPhoneOS.platform\/Library\/Developer\/CoreSimulator\/Profiles\/DeviceTypes\/iPhone 8.simdevicetype",
+  //       "name" : "iPhone 8",
+  //       "identifier" : "com.apple.CoreSimulator.SimDeviceType.iPhone-8",
+  //       "productFamily" : "iPhone"
+  //     }
+  //   ]
+  // },
+  final Map<String, Object?> _data;
+
+  String? get bundlePath => _data['bundlePath']?.toString();
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/ios/simulators.dart:701]: Instead, can you make all of these final fields, and instead do the JSON unmarshalling in a factory? This way if in a future Xcode version they--for example--rename a field, we can have multiple factories.

File: packages/flutter_tools/lib/src/macos/xcode.dart [hunk 2]
```diff
@@ -198,6 +202,19 @@ class Xcode {
...
+  Future<Version?> sdkPlatformVersion(EnvironmentType environmentType) async {
+    final RunResult runResult = await _processUtils.run(
+      <String>[...xcrunCommand(), '--sdk', getSDKNameForIOSEnvironmentType(environmentType), '--show-sdk-platform-version'],
+    );
+    if (runResult.exitCode != 0) {
+      _logger.printError('Could not find SDK Platform Version: ${runResult.stderr}');
+      return null;
+    }
+    final String versionString = runResult.stdout.trim();
+    return Version.parse(versionString);
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/macos/xcode.dart:216]: might parsing fail if there's extra STDOUT?

File: packages/flutter_tools/lib/src/macos/xcode_validator.dart [hunk 3]
```diff
@@ -68,4 +87,50 @@ class XcodeValidator extends DoctorValidator {
...
+    // Verify there is a simulator runtime installed matching the
+    // iphonesimulator SDK major version.
+    IOSSimulatorRuntime? matchingRuntime;
+    try {
+      matchingRuntime = runtimes.firstWhere(
+        (IOSSimulatorRuntime runtime) =>
+            runtime.version?.major == platformSDKVersion.major,
+      );
+    } on StateError {
+      matchingRuntime = null;
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/macos/xcode_validator.dart:127]: nit, you could return the value from line 131 here.

File: packages/flutter_tools/lib/src/macos/xcode.dart (hunk 2 above)
Comment (vashworth) [reply packages/flutter_tools/lib/src/macos/xcode.dart:216]: Can you explain what you mean by extra STDOUT? Like if it has multiple lines?

Comment (vashworth) [reply packages/flutter_tools/lib/src/macos/xcode.dart:216]: I wasn't quite sure what you meant, so I added some tests of potential things:
https://github.com/flutter/flutter/blob/62e0e9cbbed14a60be2f1f935fa1541db38089f7/packages/flutter_tools/test/general.shard/macos/xcode_test.dart#L295-L323

Comment (christopherfujino) [reply packages/flutter_tools/lib/src/macos/xcode.dart:216]: I guess I was thinking of, in a future version of `Xcode`, the output of this looks like: `v1.2.3` and `Version.parse()` fails to parse it. I suppose in that case we'd return `null` here and we'd likely find out pretty quickly, so I think this is fine.

**Review by christopherfujino — APPROVED** [review]
LGTM

**Comment by vashworth** [comment]
Oof I did something weird. I'm just going to start over at this point. 🙃

**Pull request closed without merging**